---
title: "Create a new mater table"
author: "Vladimir Souza"
date: "5/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



# Introduction

The difference of this v4 script from previous version (v3? v3_old?) version is that:
* function `is_indel_method` was added.

create the master table for wtc-11 data.

compare: all deepvariant and clair3 pipelines, and gatk_sncr.

iso-seq:
  only primary alignments
  duplicates not marked

sort-read:
  vcf downloaded from allen institute
  only chromosomes chr1, chr2, ..., chrX, chrY
  intronic regions removed



### Input variables

```{r}
### methods to validate
# name of the methods to validate
METHOD_NAMES <- c("dv", "dv_s", "dv_s_fc",
                  "c3", "c3_s", "c3_s_fc", "c3_mix",
                  "gatk_s",
                  "nc_s")
# name of the dataset used with the methods to validate
METHOD_DATASET_NAME <- "isoSeq"
# VCF files
METHOD_VCF_FILES <- c(
  "~/project_2021/datasets/wtc11/methods_to_comp/deepvariant/no_sncr/deepvariant_calls_pass.vcf.gz",
  "~/project_2021/datasets/wtc11/methods_to_comp/deepvariant/yes_sncr_no_fc/deepvariant_calls_pass.vcf.gz",
  "~/project_2021/datasets/wtc11/methods_to_comp/deepvariant/deepvariant_calls_pass.vcf.gz",
  "~/project_2021/datasets/wtc11/methods_to_comp/clair3/alone/pileup_pass.vcf.gz",
  "~/project_2021/datasets/wtc11/methods_to_comp/clair3/sncr/pileup_pass.vcf.gz",
  "~/project_2021/datasets/wtc11/methods_to_comp/clair3/sncr_fc/pileup_pass.vcf.gz",
  "~/project_2021/datasets/wtc11/methods_to_comp/clair3/mix/pileup_pass_mix_nodup.recode.vcf.gz",
  "~/project_2021/datasets/wtc11/methods_to_comp/gatk/isoSeq_wtc11.recal_pass.vcf.gz",
  "/home/vbarbo/project_2021/datasets/wtc11/methods_to_comp/nanocaller/yes_sncr_no_fc/variant_calls.final.vcf.gz"
)
# BAM of the data
METHOD_BAM_FILE <- "~/project_2021/datasets/wtc11/manipulate_data/aln_s.bam"


### ground-truth
# name
TRUTH_NAME <- "allen"
# name of the dataset used to generate the ground-truth
TRUTH_DATASET_NAME <- "shortRead"
# VCF file
TRUTH_VCF_FILE <- "/home/vbarbo/project_2021/datasets/wtc11/truth/allen_institute_wtc11_shortreads_wgs/3546dc62_AH77TTBBXX_DS-229105_GCCAAT_recalibrated_subsetChromosomes_pass.vcf.gz"


### variables
MAX_DIST_FROM_SPLICE_SITE <- 20
THREADS <- 30
```



### libraries

```{r}
library(variantCallingFromIsoSeq)
library(GenomicAlignments)
library(dplyr)
```



### Create the master table

#### Initiate the master table

Get variant positions and which method could call them.

```{r}
dat <- initiate_master_table(
  METHOD_VCF_FILES[1],
  METHOD_VCF_FILES[2],
  METHOD_VCF_FILES[3],
  METHOD_VCF_FILES[4],
  METHOD_VCF_FILES[5],
  METHOD_VCF_FILES[6],
  METHOD_VCF_FILES[7],
  METHOD_VCF_FILES[8],
  METHOD_VCF_FILES[9],
  TRUTH_VCF_FILE,
  method_names=c(METHOD_NAMES, TRUTH_NAME)
)
```

#### Add QD tag from ground-truth VCF

```{r}
# dat <- add_info_tag_from_vcf(dat, "qd_allen", "QD", TRUTH_VCF_FILE)
```

#### Add dp tag for the ground truth.

```{r}
# dat <- add_info_tag_from_vcf(dat, "dp_info_allen", "DP", TRUTH_VCF_FILE)
```

#### Add QUAL tag from DeepVariant VCF

```{r}
library(vcfR)

add_qual_from_vcf <- function(input_table, col_name, vcf_file){
  vcf <- read.vcfR(vcf_file)
  qual <- vcf@fix[,colnames(vcf@fix) == "QUAL"]
  qual <- as.numeric(qual)
  qual <- data.frame(chrm=vcf@fix[,1],
                     pos=as.integer(vcf@fix[,2]),
                     qual)
  names(qual)[3] <- col_name
  
  left_join(input_table, qual)
}

dat <- add_qual_from_vcf(dat, "qual_dv", METHOD_VCF_FILES[1])
dat <- add_qual_from_vcf(dat, "qual_dvS", METHOD_VCF_FILES[2])
dat <- add_qual_from_vcf(dat, "qual_dvSFc", METHOD_VCF_FILES[3])
dat <- add_qual_from_vcf(dat, "qual_c3", METHOD_VCF_FILES[4])
dat <- add_qual_from_vcf(dat, "qual_c3S", METHOD_VCF_FILES[5])
dat <- add_qual_from_vcf(dat, "qual_c3SFc", METHOD_VCF_FILES[6])
dat <- add_qual_from_vcf(dat, "qual_c3Mix", METHOD_VCF_FILES[7])
dat <- add_qual_from_vcf(dat, "qual_gatkS", METHOD_VCF_FILES[8])
dat <- add_qual_from_vcf(dat, "qual_ncS", METHOD_VCF_FILES[9])
```


#### get splice site positions from the BAM file and add columns about splice sites

```{r}
### load object splice_sites
load("/home/vbarbo/load_later/wtc11/wtc11_isoseq_splice_sites.RData")

dat <- add_splice_site_info_to_master_table(
  dat,
  splice_sites,
  MAX_DIST_FROM_SPLICE_SITE
)
```


#### Add the read coverage (from the BAM file) of each variant

Add iso-seq read coverage.

```{r}
### load objects `isoSeq_wtc11_cover` object
load("~/load_later/wtc11/isoSeq_wtc11_cover.RData")
method_coverage <- isoSeq_wtc11_cover

dat <- add_read_coverage_from_bam_to_master_table(
  dat,
  method_coverage,
  METHOD_DATASET_NAME
)
```

Add short-read coverage (ground truth).

```{r}
### the code to create `shortRead_coverage_1` and `shortRead_coverage_2` is in "/home/vbarbo/project_2021/scripts/wtc11/wtc11_master_table_v1.Rmd"

### ther is no `shortRead_coverage` object, because it's so big. instead i had to save the coverage of wtc-11 chromosomes separated into two objects: load shortRead_coverage_1 and shortRead_coverage_2. load these objects:
load("~/load_later/wtc11/shortRead_coverage_1_2.RData")

### add short-read coverage to master table. must do it in two steps: firstly the chromosomes in shortRead_coverage_1 and lately the ones in shortRead_coverage_2
dat_with_cover <- dat
dat_with_cover_1 <- filter(dat_with_cover, chrm %in% names(shortRead_coverage_1))
dat_with_cover_2 <- filter(dat_with_cover, chrm %in% names(shortRead_coverage_2))

dat_with_cover_1 <- add_read_coverage_from_bam_to_master_table(
  dat_with_cover_1,
  shortRead_coverage_1,
  TRUTH_DATASET_NAME
)

dat_with_cover_2 <- add_read_coverage_from_bam_to_master_table(
  dat_with_cover_2,
  shortRead_coverage_2,
  TRUTH_DATASET_NAME
)

dat_with_cover_1_2 <- rbind(dat_with_cover_1, dat_with_cover_2) %>% 
  arrange(chrm, pos)

dat <- dat_with_cover_1_2
```


#### Add the number of N-cigar reads per site

```{r}
method_bam <- readGAlignments(METHOD_BAM_FILE)

dat <- add_number_of_n_cigar_reads_to_master_table(
  dat, 
  method_bam, 
  METHOD_DATASET_NAME
)
```


#### Add column to comprare two methods

```{r}
# dat <- add_two_method_comparison_to_master_table(
#   dat, 
#   METHOD_NAMES[1],
#   METHOD_NAMES[3]
# )
# 
# dat <- add_two_method_comparison_to_master_table(
#   dat, 
#   METHOD_NAMES[2],
#   METHOD_NAMES[3]
# )
# 
# dat <- add_two_method_comparison_to_master_table(
#   dat, 
#   METHOD_NAMES[3],
#   METHOD_NAMES[4]
# )
```


#### Add column to classify the method calls (compare to the ground-truth)

```{r}
### method 1
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[1],
  TRUTH_NAME
)
### method 2
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[2],
  TRUTH_NAME
)
### method 3
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[3],
  TRUTH_NAME
)
### method 4
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[4],
  TRUTH_NAME
)
### method 5
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[5],
  TRUTH_NAME
)
### method 6
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[6],
  TRUTH_NAME
)
### method 7
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[7],
  TRUTH_NAME
)
### method 8
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[8],
  TRUTH_NAME
)
### method 9
dat <- add_method_vs_truth_comparison_to_master_table(
  dat,
  METHOD_NAMES[9],
  TRUTH_NAME
)
```


#### Add columns to inform the variant type according each method and the ground truth

For the ground-truth

```{r}
### ground truth
dat <- add_variant_type_to_master_table(dat, TRUTH_VCF_FILE, TRUTH_NAME)

### methods
dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[1],
                                        METHOD_NAMES[1])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[2],
                                        METHOD_NAMES[2])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[3],
                                        METHOD_NAMES[3])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[4],
                                        METHOD_NAMES[4])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[5],
                                        METHOD_NAMES[5])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[6],
                                        METHOD_NAMES[6])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[7],
                                        METHOD_NAMES[7])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[8],
                                        METHOD_NAMES[8])

dat <- add_variant_type_to_master_table(dat, METHOD_VCF_FILES[9],
                                        METHOD_NAMES[9])
```





#### Add columns to inform whether the variant is a indel or not (-1, 0, 1, or NA)

* 1 = is indel
* 0 = is not indel
* -1 = non defined (heterozygous alternative) (truth's gt has priority over the method's)
* NA = variant not calls by the method or the truth

```{r}
### truth
dat <- is_indel_method(dat, "allen")

### dv
dat <- is_indel_method(dat, "allen", "dv")

### dv_s
dat <- is_indel_method(dat, "allen", "dv_s")

### dv_s_fc
dat <- is_indel_method(dat, "allen", "dv_s_fc")

### c3
dat <- is_indel_method(dat, "allen", "c3")

### c3_s
dat <- is_indel_method(dat, "allen", "c3_s")

### c3_s_fc
dat <- is_indel_method(dat, "allen", "c3_s_fc")

### c3_mix
dat <- is_indel_method(dat, "allen", "c3_mix")

### gatk_s
dat <- is_indel_method(dat, "allen", "gatk_s")

### nc_s
dat <- is_indel_method(dat, "allen", "nc_s")
```


#### Add variant/sequencing_error dencity per variant

```{r}
### TRUTH_NAME
dat <- add_variant_density_of_a_method(dat, 201, TRUTH_NAME)

### the best METHOD_NAMES
dat <- add_variant_density_of_a_method(dat, 201, METHOD_NAMES[c(3,7,8)])

### METHOD_NAMES
# dat <- add_variant_density_of_a_method(dat, 201, METHOD_NAMES)
```


#### add genotype

```{r}
dat <- add_format_tag_from_vcf(dat, "allen", "GT", TRUTH_VCF_FILE)

dat <- add_format_tag_from_vcf(dat, "dv_s_fc", "GT", METHOD_VCF_FILES[3])

dat <- add_format_tag_from_vcf(dat, "c3_mix", "GT", METHOD_VCF_FILES[7])

dat <- add_format_tag_from_vcf(dat, "gatk_s", "GT", METHOD_VCF_FILES[8])

dat <- add_format_tag_from_vcf(dat, "nc_s", "GT", METHOD_VCF_FILES[9])

k <- is.na( dat[ ,c("gt_allen", "gt_dv_s_fc", "gt_c3_mix", "gt_gatk_s", "gt_nc_s") ] )
dat[ ,c("gt_allen", "gt_dv_s_fc", "gt_c3_mix", "gt_gatk_s", "gt_nc_s") ] [k] <- "0/0"
```


#### add homopolimer length (from reference FASTA), but only for 0/1 and 1/1 genotypes and indels

```{r}
# library(Biostrings)
# genome_ref_file <- "/home/vbarbo/project_2021/datasets/reference/GRCh38.p13_genome_only_chrm/GRCh38.p13_all_chr.fasta"
# genome_ref <- readDNAStringSet(genome_ref_file)
# library(sarlacc)
# homopolymers <- homopolymerFinder(genome_ref)
load("/home/vbarbo/load_later/homopolymers_GRCh38.p13.RData") # homopolymers object
homopolymers_bk <- homopolymers

dat <- add_homopolymer_length_when_indels(dat, homopolymers)
```


### Take a look at the master table

```{r}
head(dat)
dat_bk <- dat
```


### Save master table to a file

```{r}
mt_wtc11_allMethods_notFiltered <- dat
save(mt_wtc11_allMethods_notFiltered,
     file="/home/vbarbo/project_2021/projects/iso-seq_variant_calling/scripts_for_all_analyses_in_the_paper/3_creating_master_table/wct11/mt_wtc11_allMethods_notFiltered_v6.RData",
     compress="xz")
```


### Filtering master table

* short-read variant density: x <= 3                          (ignore positions)
* ignore intronic regions (isoSeq_coverage > 0)               (ignore positions)
* DeepVariant QUAL: x >= 15                                   (remove method calling)
* short-read coverare : quantil 5% <= x <= 95%                (ignore positions)

```{r}
# no filtering
dat1 <- dat
dim(dat1)

shortread_cover_quantiles <- quantile(dat1$shortRead_coverage, probs=c(.05, .95))

# variant/sequence_error density -- filter using the ground truth
k <- !( dat1$variantDensity_allen > 3 )
dat1 <- dat1[k,]
dim(dat1)

# ignore intronic regions (iso-seq coverage equal to zero)
dat1 <- filter(dat1, isoSeq_coverage>0)
dim(dat1)

# ignore regions based on short-read coverage quantiles: keep 5% <= x <= 95%
dat1 <- filter(dat1, shortRead_coverage >= shortread_cover_quantiles[1] &
                 shortRead_coverage <= shortread_cover_quantiles[2])
dim(dat1)

# Filter out DeepVariant calls by QUAL -- but keep the position
# dv
k <- which( dat1$qual_dv < 15 )
dat1$in_dv[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[1],
  TRUTH_NAME,
  replace_column=TRUE
)
# dv_s
k <- which( dat1$qual_dvS < 15 )
dat1$in_dv_s[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[2],
  TRUTH_NAME,
  replace_column=TRUE
)
# dv_s_fc
k <- which( dat1$qual_dvSFc < 15 )
dat1$in_dv_s_fc[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[3],
  TRUTH_NAME,
  replace_column=TRUE
)


########## don't filter by Clair3 QUAL

########## don't filter by NanoCaller QUAL
```

save filtered master table

```{r}
mt_wtc11_allMethods_filtered <- dat1
save(mt_wtc11_allMethods_filtered,
     file="/home/vbarbo/project_2021/projects/iso-seq_variant_calling/scripts_for_all_analyses_in_the_paper/3_creating_master_table/wct11/mt_wtc11_allMethods_filtered_v6.RData",
     compress="xz")
```

also, save filtered master table using iso-seq data to filter high variant density regions:  
use variants calls from sncr+fc+deepvariant, cair3 mix, and sncr+gatk.

```{r}
dat1 <- dat
shortread_cover_quantiles <- quantile(dat1$shortRead_coverage, probs=c(.05, .95))

# >>> variant/sequence_error density -- filter using the ground truth and iso-seq (all methods)
k <- !( dat1$variantDensity_allen > 3 )
k1 <- !( dat1$variantDensity_dvSFc_c3Mix_gatkS > 3 )
k <- k & k1
dat1 <- dat1[k,]
dim(dat1)
# <<<

dat1 <- filter(dat1, isoSeq_coverage>0)
dim(dat1)
dat1 <- filter(dat1, shortRead_coverage >= shortread_cover_quantiles[1] &
                 shortRead_coverage <= shortread_cover_quantiles[2])
k <- which( dat1$qual_dv < 15 )
dat1$in_dv[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[1],
  TRUTH_NAME,
  replace_column=TRUE
)
k <- which( dat1$qual_dvS < 15 )
dat1$in_dv_s[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[2],
  TRUTH_NAME,
  replace_column=TRUE
)
k <- which( dat1$qual_dvSFc < 15 )
dat1$in_dv_s_fc[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[3],
  TRUTH_NAME,
  replace_column=TRUE
)


### variant density given by all 8 methods used for filtering
mt_wtc11_allMethods_filtered_filterVariantDensityIsoSeqDvsfcC3mixGatks <- dat1
save(mt_wtc11_allMethods_filtered_filterVariantDensityIsoSeqDvsfcC3mixGatks,
     file="~/load_later/homopolymer_analysis/mt_wtc11_allMethods_filtered_v5_filterVariantDensityIsoSeqDvsfcC3mixGatks.RData")
```


once again, save filtered master table using iso-seq data to filter high variant density regions:  
use variants calls from all methods

```{r}
dat1 <- dat
shortread_cover_quantiles <- quantile(dat1$shortRead_coverage, probs=c(.05, .95))

# >>> variant/sequence_error density -- filter using the ground truth and iso-seq (all methods)
k <- !( dat1$variantDensity_allen > 3 )
k1 <- !( dat1$variantDensity_dv_dvS_dvSFc_c3_c3S_c3SFc_c3Mix_gatkS > 3 )
k <- k & k1
dat1 <- dat1[k,]
dim(dat1)
# <<<

dat1 <- filter(dat1, isoSeq_coverage>0)
dim(dat1)
dat1 <- filter(dat1, shortRead_coverage >= shortread_cover_quantiles[1] &
                 shortRead_coverage <= shortread_cover_quantiles[2])
k <- which( dat1$qual_dv < 15 )
dat1$in_dv[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[1],
  TRUTH_NAME,
  replace_column=TRUE
)
k <- which( dat1$qual_dvS < 15 )
dat1$in_dv_s[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[2],
  TRUTH_NAME,
  replace_column=TRUE
)
k <- which( dat1$qual_dvSFc < 15 )
dat1$in_dv_s_fc[k] <- 0
dat1 <- add_method_vs_truth_comparison_to_master_table(
  dat1,
  METHOD_NAMES[3],
  TRUTH_NAME,
  replace_column=TRUE
)


### variant density given by all 8 methods used for filtering
mt_wtc11_allMethods_filtered_filterVariantDensityIsoSeqAllMethods <- dat1
save(mt_wtc11_allMethods_filtered_filterVariantDensityIsoSeqAllMethods,
     file="~/load_later/homopolymer_analysis/mt_wtc11_allMethods_filtered_v5_filterVariantDensityIsoSeqAllMethods.RData")
```


